File: bench/io/bench_readlines.c
/* Count lines and bytes by reading one line at a time.
 *
 * The point is the line loop, not the parsing: each language uses whatever its
 * standard library offers for iterating a text file line by line, which is what
 * a program processing a log or a CSV would actually write.
 *
 * result folds both counts into one integer so a miscount in either shows up:
 *   lines * 100000000 + bytes   (the generator keeps bytes below 100 million) */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(void)
{
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    FILE *f = fopen("io_input.txt", "rb");
    if (!f) { fprintf(stderr, "cannot open io_input.txt\n"); return 1; }

    char *line = NULL;
    size_t cap = 0;
    ssize_t n;
    long long lines = 0, bytes = 0;
    while ((n = getline(&line, &cap, f)) > 0) { lines++; bytes += n; }
    free(line);
    fclose(f);

    clock_gettime(CLOCK_MONOTONIC, &t1);
    printf("result: %lld\n", lines * 100000000LL + bytes);
    printf("elapsed: %lld\n",
           (long long)((t1.tv_sec - t0.tv_sec) * 1000 + (t1.tv_nsec - t0.tv_nsec) / 1000000));
    return 0;
}

File: bench/io/bench_readlines.rs
// Count lines and bytes, one line at a time. See bench_readlines.c for the
// result fold.
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;

fn main() {
    let t0 = Instant::now();
    let f = File::open("io_input.txt").expect("cannot open io_input.txt");
    let mut r = BufReader::new(f);
    let mut buf = String::new();
    let mut lines: i64 = 0;
    let mut total: i64 = 0;
    loop {
        buf.clear();
        match r.read_line(&mut buf) {
            Ok(0) => break,
            Ok(n) => {
                lines += 1;
                total += n as i64;
            }
            Err(_) => break,
        }
    }
    let ms = t0.elapsed().as_millis();
    println!("result: {}", lines * 100000000 + total);
    println!("elapsed: {}", ms);
}
